File: src/main/java/com/warren/bt/Decorator.java
package com.warren.bt;

public abstract class Decorator implements Node {
  protected Node child;

  public Decorator(Node child) {
    this.child = child;
  }

  public Decorator() {
    this(null);
  }

  public Decorator setChild(Node child) {
    this.child = child;
    return this;
  }

  @Override
  public void reset() {
    if (child != null) {
      child.reset();
    }
  }
}

File: src/main/kotlin/com/warren/bt/Decorator.kt
package com.warren.bt

abstract class Decorator(override val name: String = "Decorator", var child: Node? = null) :
        Node(name) {
  fun assignChild(node: Node) {
    child = node
  }

  override fun reset() {
    child?.reset()
    onReset()
  }

  protected open fun onReset() {}
}
